pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix

```diff
@@ -41,6 +41,9 @@ rustPlatform.buildRustPackage rec {
 
   cargoSha256 = "sha256-edi6/Md6KebKM3wHArZe1htUCg0/BqMVZKA4xEH25GI=";
 
+  # lld: error: unknown argument '-Wl,--undefined=AUDITABLE_VERSION_INFO'
+  auditable = false;
```

Comment:
I think we should have something that takes into account the `platforms` and disable `auditable` automatically based on this.

Reply:
Maybe this isn't a problem with cross-compiling as this error hints at a problem of cargo-auditable in combination with ldd.

Reply:
Hold on, I think this is this PR's problem. You should probably pass clang as a linker to Rust rather than lld directly.

Reply:
Sorry, that's not working out, `-nostartfiles` doesn't work as desired here.
```patch
diff --git pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix
index 42cc3ab07a8..24adc5d1fc0 100644
--- pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix
+++ pkgs/applications/virtualization/rust-hypervisor-firmware/default.nix
@@ -3,6 +3,7 @@
 , makeRustPlatform
 , hostPlatform
 , targetPlatform
+, clang
 , lld
 }:

@@ -44,10 +45,12 @@ rustPlatform.buildRustPackage rec {
   RUSTC_BOOTSTRAP = 1;

   nativeBuildInputs = [
-    lld
+    clang lld
   ];

-  RUSTFLAGS = "-C linker=lld -C linker-flavor=ld.lld";
+  RUSTFLAGS = "-C linker=clang -C link-arg=-nostartfiles";

   # Tests don't work for `no_std`. See https://os.phil-opp.com/testing/
   doCheck = false;
```

The upstream rust-hypervisor-firmware project hardcoded `ld.lld`. Perhaps cargo-auditable should be more sensitive to the `linker-flavor`?

Reply:
cc @Shnatsel

Reply:
Okay, this is a firmware file and I don't think there's anything that can extract the info from those yet, so it's reasonable to disable audit info here and for `*-unknown-none` platforms, at least for now.

That said, if you want to investigate...

The argument  coming from [this line](https://github.com/rust-secure-code/cargo-auditable/blob/47121426a859ef14a91ba9be0bce19d4d0019faf/cargo-auditable/src/rustc_wrapper.rs#L56) in `cargo auditable`. It does indeed expect the linker to be `clang` or similar. `rustc` uses the host's C compiler as a linker wrapper on Unix.

Could you try changing that line to `command.arg("-Clink-arg=--undefined=AUDITABLE_VERSION_INFO");` (without the `-Wl`) and see if that fixes it?

Alternatively you can point me to the upstream project and I'll investigate if that builds with `cargo auditable`.

Reply:
Here is the upstream project: https://github.com/cloud-hypervisor/rust-hypervisor-firmware

Reply:
Thanks, I'll take a look tomorrow.

Reply:
I've investigated and it turns out that the upstream project uses _both_ naked LLD and LLD via a compiler wrapper, which effectively breaks the `-C link-arg=` rustc flag that `cargo auditable` relies on.

If I use `"-Clink-arg=-Wl,--undefined=AUDITABLE_VERSION_INFO"` like `cargo auditable` currently does I get the error you've seen:

`rust-lld: error: unknown argument '-Wl,--undefined=AUDITABLE_VERSION_INFO'`

but dropping the `-Wl,` gives errors from `cc` instead:

`cc: error: unrecognized command-line option '--undefined=AUDITABLE_VERSION_INFO'`

So there's no way to pass `-C link-arg=` and have it work in that project.

Reply:
Thanks for the investigation! Since this is specific to the project, I think we can just proceed with this fix, what do you think @RaitoBezarius?

Reply:
If it's not too much, I would appreciate if we opened an issue upstream and linked to it in the comments.

Reply:
Link added